govcd/nsxv_distributed_firewall.go

```diff
@@ -0,0 +1,501 @@
+/*
+ * Copyright 2023 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+)
+
+// NsxvDistributedFirewall defines a distributed firewall for a NSX-V VDC
+type NsxvDistributedFirewall struct {
+	VdcId         string                       // The ID of the VDC
+	Configuration *types.FirewallConfiguration // The latest firewall configuration
+	Etag          string
+	enabled       bool    // internal flag that signifies whether the firewall is enabled
+	client        *Client // internal usage client
+
+	// Attributes below to be removed if we decide that we don't need services
+	Services      []types.Application      // The list of services for this VDC
+	ServiceGroups []types.ApplicationGroup // The list of service groups for this VDC
+}
+
+// Protocols
+const (
+	DFWProtocolTcp  = "TCP"
+	DFWProtocolUdp  = "UDP"
+	DFWProtocolIcmp = "ICMP"
+)
+
+// Action types
+const (
+	DFWActionAllow = "allow"
+	DFWActionDeny  = "deny"
+)
+
+// Directions
+const (
+	DFWDirectionIn    = "in"
+	DFWDirectionOut   = "out"
+	DFWDirectionInout = "inout"
+)
+
+// Types of packet
+const (
+	DFWPacketAny  = "any"
+	DFWPacketIpv4 = "ipv4"
+	DFWPacketIpv6 = "ipv6"
+)
+
+// Elements of Source, Destination, and Applies-To
+const (
+	DFWElementVdc            = "VDC"
+	DFWElementVirtualMachine = "VirtualMachine"
+	DFWElementNetwork        = "Network"
+	DFWElementEdge           = "Edge"
+	DFWElementIpSet          = "IPSet"
+	DFWElementIpv4           = "Ipv4Address"
+)
+
+// Types of service
+const (
+	DFWServiceTypeApplication      = "Application"
+	DFWServiceTypeApplicationGroup = "ApplicationGroup"
+)
+
+var NsxvProtocolCodes = map[string]int{
+	DFWProtocolTcp:  6,
+	DFWProtocolUdp:  17,
+	DFWProtocolIcmp: 1,
+}
+
+// NewNsxvDistributedFirewall creates a new NsxvDistributedFirewall
+func NewNsxvDistributedFirewall(client *Client, vdcId string) *NsxvDistributedFirewall {
+	return &NsxvDistributedFirewall{
+		client: client,
+		VdcId:  extractUuid(vdcId),
+	}
+}
+
+// buildUrl uses the Client base URL to create a customised URL
+func (client *Client) buildUrl(elements ...string) (string, error) {
```

Comment:
Could this function maybe be moved to `api.go` as it seems universal and could be used in other files.

Reply:
moved